src/rust/catnap/win/winsock.rs

```diff
@@ -0,0 +1,348 @@
+    /// Safety: the requested socket option must agree with the ABI of T.
+    #[allow(unused)]
+    pub unsafe fn setsockopt<'a, T>(&self, s: SOCKET, level: i32, opt: i32, val: Option<&'a T>) -> Result<(), Fail> {
+        Self::do_setsockopt(s, level, opt, val)
+    }
+
+    /// Implementation of getsockopt.
+    pub(super) unsafe fn do_getsockopt<T>(s: SOCKET, level: i32, optname: i32) -> Result<T, Fail> {
+        let mut out: MaybeUninit<T> = MaybeUninit::zeroed();
+        let optval: PSTR = PSTR::from_raw(out.as_mut_ptr().cast());
+        let mut optlen: i32 =
+            i32::try_from(std::mem::size_of::<T>()).map_err(|_| Fail::new(libc::E2BIG, "option type too large"))?;
+        if unsafe { getsockopt(s, level, optname, optval, &mut optlen) } == 0 {
+            Ok(unsafe { out.assume_init() })
+        } else {
+            Err(expect_last_wsa_error())
+        }
+    }
+
+    /// Get a socket option (via getsockopt) and return the option as a structure type.
+    /// Safety: the requested socket option must initialize a value of type T.
+    pub unsafe fn getsockopt<T>(&self, s: SOCKET, level: i32, optname: i32) -> Result<T, Fail> {
+        Self::do_getsockopt(s, level, optname)
+    }
+
+    /// Get or initialize a new `SocketExtensions` instance for a  socket. Extensions are stored by socket provider,
+    /// which may be shared by multiple sockets.
+    fn get_or_init_extensions(&mut self, s: SOCKET) -> Result<Rc<SocketExtensions>, Fail> {
+        let protocol: WSAPROTOCOL_INFOW = unsafe { self.getsockopt(s, SOL_SOCKET, SO_PROTOCOL_INFOW) }?;
+
+        let extensions: &mut Weak<SocketExtensions> = self
+            .extensions_by_provider
+            .entry(protocol.ProviderId)
+            .or_insert(Weak::default());
+        if let Some(extensions) = extensions.upgrade() {
+            return Ok(extensions);
+        }
+
+        let new_extensions: Rc<SocketExtensions> = SocketExtensions::new(s)?;
+        *extensions = Rc::downgrade(&new_extensions);
+
+        Ok(new_extensions)
+    }
+
+    /// Create a raw Winsock socket.
+    pub(super) fn raw_socket(
+        domain: libc::c_int,
+        typ: libc::c_int,
+        protocol: libc::c_int,
+        protocol_info: Option<&WSAPROTOCOL_INFOW>,
+        flags: u32,
+    ) -> Result<SOCKET, Fail> {
+        let protocol_info: Option<*const WSAPROTOCOL_INFOW> = protocol_info.map(|i| i as *const WSAPROTOCOL_INFOW);
+        match unsafe { WSASocketW(domain, typ, protocol, protocol_info, 0, flags) } {
+            INVALID_SOCKET => Err(expect_last_wsa_error()),
+            socket => Ok(socket),
+        }
+    }
+
+    /// Create a new socket.
+    pub fn socket(
+        &mut self,
+        domain: libc::c_int,
+        typ: libc::c_int,
+        protocol: libc::c_int,
+        config: &WinConfig,
+        iocp: &IoCompletionPort,
+    ) -> Result<Socket, Fail> {
+        // Safety: the socket handle returned from WSASocketW must be closed to clean up resources.
```

Comment:
Could you clarify what this comment means? Do you mean it must be cleaned up on close()?